src/pdb/string.rs

```diff
@@ -0,0 +1,271 @@
+                    content: LongBody::Ucs2le(string.encode_utf16().collect()),
+                }))
+            }
+        } else {
+            Err(StringError::TooLong)
+        }
+    }
+    /// Creates a DeviceSQLString containing an ISRC instead of an expected string
+    ///
+    /// The DeviceSQL Database as used by rekordbox has a strange oddity
+    /// where it serializes strings containing a tracks ISRC (International
+    /// Standard Recording Code) in an unexpected format, if this is desired,
+    /// use this constructor function instead of [DeviceSQLString::new]
+    pub fn new_isrc(string: String) -> Result<Self, StringError> {
+        if string.is_empty() {
+            return Ok(Self::empty());
+        }
+        // Basic validation taken from:
+        // https://isrc.ifpi.org/downloads/ISRC_Bulletin-2015-01.pdf
+        if string.len() != 12 || !string.is_ascii() {
+            return Err(StringError::InvalidISRC);
+        }
+        Ok(Self(DeviceSQLStringImpl::Long {
+            content: LongBody::Isrc(NullString::from_string(string)),
+        }))
+    }
+    /// Extract the Rust string from the DeviceSQLString.
+    ///
+    /// Consumes itself in the process.
+    pub fn into_string(self) -> Result<String, StringError> {
+        match self.0 {
+            DeviceSQLStringImpl::ShortASCII { content: vec, .. }
+            | DeviceSQLStringImpl::Long {
+                content: LongBody::Ascii(vec),
+                ..
+            } => String::from_utf8(vec).map_err(|_| StringError::Encoding),
+            DeviceSQLStringImpl::Long {
+                content: LongBody::Isrc(str),
+                ..
+            } => str
+                .into_string_lossless()
+                .map_err(|_| StringError::Encoding),
+            DeviceSQLStringImpl::Long {
+                content: LongBody::Ucs2le(vec),
+            } => String::from_utf16(&vec).map_err(|_| StringError::Encoding),
+        }
+    }
+    /// Create an empty DeviceSQLString
+    ///
+    /// Should be used to construct known empty strings.
+    #[must_use]
+    pub const fn empty() -> Self {
+        Self(DeviceSQLStringImpl::ShortASCII {
+            content: Vec::new(),
+        })
+    }
+}
+
+/// A String encapsulating how DeviceSQL VARCHAR's are structured
+///
+/// This implementation forces them to be immutable for now
+#[binrw]
+#[derive(Debug, PartialEq)]
+enum DeviceSQLStringImpl {
+    /// Short-String optimization case
+    ShortASCII {
+        #[br(temp, assert(header & 0b1 == 1))]
+        #[bw(calc = (((content.len() + 1) << 1) | 1) as u8)]
+        header: u8,
+
+        #[br(count = (header >> 1) - 1)]
+        content: Vec<u8>,
+    },
+    /// Regular long form strings, containing possibly different encodings
+    Long {
+        #[br(temp)]
+        #[br(assert(flags & 0b1 == 0))]
+        #[bw(calc = content.flags())]
+        flags: u8,
+
+        #[br(temp)]
+        #[bw(calc = content.byte_count().unwrap() + 4)]
```

Comment:
This may panic... is there no way to return an error? I'm not sure if this is possible.

Reply:
Its only going to return an error if the string we constructed was already too big in the first place (which we already check on construction). I also opted for this approach because I could not find another way of making this panic-less.